### ISSUE TITLE
ugly slurs on grace notes [regression]

### ISSUE THREAD

**Issue by frakel** [rism-digital/verovio#2622]
Apparently the shape of slurs on grace notes has deteriorated starting with Version 3.7.x.
See the current rendering of
<details><summary>this example</summary>
<p>

```xml
<?xml version="1.0" encoding="UTF-8"?>
<?xml-model href="https://music-encoding.org/schema/4.0.1/mei-CMN.rng" type="application/xml" schematypens="http://relaxng.org/ns/structure/1.0"?>
<?xml-model href="https://music-encoding.org/schema/4.0.1/mei-CMN.rng" type="application/xml" schematypens="http://purl.oclc.org/dsdl/schematron"?>
<mei meiversion="4.0.0" xmlns="http://www.music-encoding.org/ns/mei">
  <meiHead>
    <fileDesc>
      <titleStmt>
        <title label="NMA digital" type="unit">slurs on grace notes</title>
        <contributor>DIME/FK</contributor>
      </titleStmt>
      <pubStmt/>
    </fileDesc>
  </meiHead>
  <music>
    <body>
      <mdiv>
        <score>
          <scoreDef meter.count="4" meter.sym="common" meter.unit="4">
            <staffGrp>
              <staffDef clef.line="2" clef.shape="G" lines="5" n="1" xml:id="sd1"/>
            </staffGrp>
          </scoreDef>
          <section>
            <measure xml:id="k282m1" n="31">
              <staff n="1">
                <layer n="1" xml:id="layer_12666">
                  <beam xml:id="beam_12720">
                    <note dur="16" oct="4" pname="a" xml:id="note_12726"/>
                    <note dur="32" grace="unknown" oct="4" pname="g" xml:id="note_12732"/>
                    <note dur="16" oct="4" pname="b" xml:id="note_12738"/>
                    <note dur="32" grace="unknown" oct="4" pname="a" xml:id="note_12744"/>
                    <note dur="16" oct="5" pname="c" xml:id="note_12750"/>
                    <note dur="32" grace="unknown" oct="4" pname="b" xml:id="note_12756"/>
                    <note dur="16" oct="5" pname="d" xml:id="note_12762"/>
                  </beam>
                </layer>
              </staff>
              <slur color="red" curvedir="below" endid="#note_12738" staff="1" startid="#note_12732" xml:id="slur_13044"/>
              <slur color="red" curvedir="below" endid="#note_12750" staff="1" startid="#note_12744" xml:id="slur_13050"/>
              <slur color="red" curvedir="below" endid="#note_12762" staff="1" startid="#note_12756" xml:id="slur_13056"/>
            </measure>
          </section>
        </score>
      </mdiv>
    </body>
  </music>
</mei>

```
</details>
compared with version 3.6 where the rendering of the example above was
ok:

![image](https://user-images.githubusercontent.com/5627140/152782761-28f75008-6c52-4e50-81e2-19b407cdb116.png)

**Comment by frakel** [comment]
I just noticed that the slurs continuously change their shape depending on their pitch. Bevore 3.7 they stayed way better rendered in all positions:
![image](https://user-images.githubusercontent.com/5627140/152808353-b29ab156-a62f-49b8-af78-a0caff30cbc0.png)
<details><summary>see the code</summary>
<p>

```xml
<?xml version="1.0" encoding="UTF-8"?>
<?xml-model href="https://music-encoding.org/schema/4.0.1/mei-CMN.rng" type="application/xml" schematypens="http://relaxng.org/ns/structure/1.0"?>
<?xml-model href="https://music-encoding.org/schema/4.0.1/mei-CMN.rng" type="application/xml" schematypens="http://purl.oclc.org/dsdl/schematron"?>
<mei meiversion="4.0.0" xmlns="http://www.music-encoding.org/ns/mei">
  <meiHead>
    <fileDesc>
      <titleStmt>
        <title label="NMA digital" type="unit">slurs on grace notes</title>
        <contributor>DIME/FK</contributor>
      </titleStmt>
      <pubStmt/>
    </fileDesc>
  </meiHead>
  <music>
    <body>
      <mdiv>
        <score>
          <scoreDef meter.count="4" meter.sym="common" meter.unit="4">
            <staffGrp>
              <staffDef clef.line="2" clef.shape="G" lines="5" n="1" xml:id="sd1"/>
            </staffGrp>
          </scoreDef>
          <section>
            <measure n="31" xml:id="k282m1">
              <staff n="1">
                <layer n="1" xml:id="layer_12666">
                  <beam xml:id="beam_12720">
                    <note dur="16" oct="4" pname="a" xml:id="note_12726"/>
                    <note dur="32" grace="unknown" oct="4" pname="g" xml:id="note_12732"/>
                    <note dur="16" oct="4" pname="b" xml:id="note_12738"/>
                    <note dur="32" grace="unknown" oct="4" pname="a" xml:id="note_12744"/>
                    <note dur="16" oct="5" pname="c" xml:id="note_12750"/>
                    <note dur="32" grace="unknown" oct="4" pname="b" xml:id="note_12756"/>
                    <note dur="16" oct="5" pname="d" xml:id="note_12762"/>
                  </beam>
                  <note dur="32" grace="unknown" oct="5" pname="f" xml:id="note_12732b"/>
                  <beam xml:id="beam_12720a">
                    <note dur="16" oct="5" pname="a" xml:id="note_12726a"/>
                    <note dur="32" grace="unknown" oct="5" pname="g" xml:id="note_12732a"/>
                    <note dur="16" oct="5" pname="b" xml:id="note_12738a"/>
                    <note dur="32" grace="unknown" oct="5" pname="a" xml:id="note_12744a"/>
                    <note dur="16" oct="6" pname="c" xml:id="note_12750a"/>
                    <note dur="32" grace="unknown" oct="5" pname="b" xml:id="note_12756a"/>
                    <note dur="16" oct="6" pname="d" xml:id="note_12762a"/>
                  </beam>
                </layer>
              </staff>
              <slur color="red" curvedir="below" endid="#note_12738" staff="1" startid="#note_12732" xml:id="slur_13044"/>
              <slur color="red" curvedir="below" endid="#note_12750" staff="1" startid="#note_12744" xml:id="slur_13050"/>
              <slur color="red" curvedir="below" endid="#note_12762" staff="1" startid="#note_12756" xml:id="slur_13056"/>
              <slur color="red" curvedir="below" endid="#note_12726a" staff="1" startid="#note_12732b" xml:id="slur_13044b"/>
              <slur color="red" curvedir="below" endid="#note_12738a" staff="1" startid="#note_12732a" xml:id="slur_13044a"/>
              <slur color="red" curvedir="below" endid="#note_12750a" staff="1" startid="#note_12744a" xml:id="slur_13050a"/>
              <slur color="red" curvedir="below" endid="#note_12762a" staff="1" startid="#note_12756a" xml:id="slur_13056a"/>
            </measure>
          </section>
        </score>
      </mdiv>
    </body>
  </music>
</mei>

```
</details>

The same happens on downwards scales, whereas here only the higher notes (on ledger lines) seem to be affected:
![image](https://user-images.githubusercontent.com/5627140/152809363-79711365-c758-4457-a530-f579efa5f00c.png)
<details><summary>see the code</summary>
<p>

```xml
<?xml version="1.0" encoding="UTF-8"?>
<?xml-model href="https://music-encoding.org/schema/4.0.1/mei-CMN.rng" type="application/xml" schematypens="http://relaxng.org/ns/structure/1.0"?>
<?xml-model href="https://music-encoding.org/schema/4.0.1/mei-CMN.rng" type="application/xml" schematypens="http://purl.oclc.org/dsdl/schematron"?>
<mei meiversion="4.0.0" xmlns="http://www.music-encoding.org/ns/mei">
  <meiHead>
    <fileDesc>
      <titleStmt>
        <title label="NMA digital" type="unit">slurs on grace notes</title>
        <contributor>DIME/FK</contributor>
      </titleStmt>
      <pubStmt/>
    </fileDesc>
  </meiHead>
  <music>
    <body>
      <mdiv>
        <score>
          <scoreDef meter.count="4" meter.sym="common" meter.unit="4">
            <staffGrp>
              <staffDef clef.line="2" clef.shape="G" lines="5" n="1" xml:id="sd1"/>
            </staffGrp>
          </scoreDef>
          <section>
            <measure xml:id="m17m20_k309_001">
              <staff n="1">
                <layer n="1" xml:id="layer_5970">
                  <note dur="16" grace="unknown" oct="6" pname="c" xml:id="note_6564a"/>
                  <beam xml:id="beam_6570a">
                    <note dur="8" oct="5" pname="b" xml:id="note_6576a"/>
                    <note dur="16" oct="5" pname="a" xml:id="note_6582a"/>
                    <note dur="16" oct="5" pname="g" xml:id="note_6588a"/>
                  </beam>
                  <note dur="16" grace="unknown" oct="5" pname="b" xml:id="note_6564"/>
                  <beam xml:id="beam_6570">
                    <note dur="8" oct="5" pname="a" xml:id="note_6576"/>
                    <note dur="16" oct="5" pname="g" xml:id="note_6582"/>
                    <note dur="16" oct="5" pname="f" xml:id="note_6588"/>
                  </beam>
                  <note dur="16" grace="unknown" oct="5" pname="g" xml:id="note_6006"/>
                  <beam xml:id="beam_6012">
                    <note dur="8" oct="5" pname="f" xml:id="note_6018"/>
                    <note dur="16" oct="5" pname="e" xml:id="note_6024"/>
                    <note dur="16" oct="5" pname="d" xml:id="note_6030"/>
                  </beam>
                  <note dur="16" grace="unknown" oct="5" pname="f" xml:id="note_6006a"/>
                  <beam xml:id="beam_6012a">
                    <note dur="8" oct="5" pname="e" xml:id="note_6018a"/>
                    <note dur="16" oct="5" pname="d" xml:id="note_6024a"/>
                    <note dur="16" oct="5" pname="c" xml:id="note_6030a"/>
                  </beam>
                </layer>
              </staff>
              <slur color="red" curvedir="below" endid="#note_6576a" staff="1" startid="#note_6564a" xml:id="slur_6690a"/>
              <slur color="red" curvedir="below" endid="#note_6576" staff="1" startid="#note_6564" xml:id="slur_6690"/>
              <slur color="red" curvedir="below" endid="#note_6018" staff="1" startid="#note_6006" xml:id="slur_6126"/>
              <slur color="red" curvedir="below" endid="#note_6018a" staff="1" startid="#note_6006a" xml:id="slur_6126a"/>
            </measure>
            <measure xml:id="m17m20_k309_001a">
              <staff n="1">
                <layer n="1" xml:id="layer_5970a">
                  <note dur="16" grace="unknown" oct="5" pname="c" xml:id="note_6564b"/>
                  <beam xml:id="beam_6570b">
                    <note dur="8" oct="4" pname="b" xml:id="note_6576b"/>
                    <note dur="16" oct="4" pname="a" xml:id="note_6582b"/>
                    <note dur="16" oct="4" pname="g" xml:id="note_6588b"/>
                  </beam>
                  <note dur="16" grace="unknown" oct="4" pname="b" xml:id="note_6564ab"/>
                  <beam xml:id="beam_6570ab">
                    <note dur="8" oct="4" pname="a" xml:id="note_6576ab"/>
                    <note dur="16" oct="4" pname="g" xml:id="note_6582ab"/>
                    <note dur="16" oct="4" pname="f" xml:id="note_6588ab"/>
                  </beam>
                  <note dur="16" grace="unknown" oct="4" pname="g" xml:id="note_6006ab"/>
                  <beam xml:id="beam_6012ab">
                    <note dur="8" oct="4" pname="f" xml:id="note_6018ab"/>
                    <note dur="16" oct="4" pname="e" xml:id="note_6024ab"/>
                    <note dur="16" oct="4" pname="d" xml:id="note_6030ab"/>
                  </beam>
                  <note dur="16" grace="unknown" oct="4" pname="f" xml:id="note_6006b"/>
                  <beam xml:id="beam_6012b">
                    <note dur="8" oct="4" pname="e" xml:id="note_6018b"/>
                    <note dur="16" oct="4" pname="d" xml:id="note_6024b"/>
                    <note dur="16" oct="4" pname="c" xml:id="note_6030b"/>
                  </beam>
                </layer>
              </staff>
              <slur color="red" curvedir="below" endid="#note_6576b" staff="1" startid="#note_6564b" xml:id="slur_6690ab"/>
              <slur color="red" curvedir="below" endid="#note_6576ab" staff="1" startid="#note_6564ab" xml:id="slur_6690b"/>
              <slur color="red" curvedir="below" endid="#note_6018ab" staff="1" startid="#note_6006ab" xml:id="slur_6126b"/>
              <slur color="red" curvedir="below" endid="#note_6018b" staff="1" startid="#note_6006b" xml:id="slur_6126bb"/>
            </measure>
          </section>
        </score>
      </mdiv>
    </body>
  </music>
</mei>

```
</details>

**Comment by frakel** [comment]
short note on the side concerning the second screenshot of my last comment:  

If the note's stem.dir is _up_ the beam suddenly is parallel with the staff lines and no longer tilted in accordance with the downwards movement of the notes; previously the beams were equally tilted independently from their children's stem.dir

![image](https://user-images.githubusercontent.com/5627140/152812728-f7d25c93-6e70-4342-b794-36f563ad2abf.png)

**Comment by frakel** [comment]
The slur endings in the first and second beam all of them oddly enough seem to point to a fictional d''

**Comment by DavidBauer1984** [comment]
The fix for #2592 did not affect grace slurs inside beams, they are treated separately in the code. But we could try similar improvements for the inside beam case as done there.
